### PR TITLE
netcoredbg: init at 1.2.0-825

### LINKED ISSUE #103938: Package netcoredbg
**Project description**
netcoredbg is an open-source debugger for .NET Core applications. The most popular one, vsdbg, is a binary blob distributed with the Visual Studio Code C# extension. However, [vsdbg's license](https://visualstudio.microsoft.com/license-terms/mt644895/) is incredibly restrictive. netcoredbg is a [drop-in replacement](https://github.com/VSCodium/vscodium/issues/82), and is also compatible with [other editors](https://github.com/puremourning/vimspector#supported-languages).

**Metadata**

* homepage URL: https://github.com/Samsung/netcoredbg
* source URL: https://github.com/Samsung/netcoredbg
* license: MIT
* platforms: Linux, macOS, Windows (possibly others?)

### PR DESCRIPTION
**Pull request by leo60228**
###### Motivation for this change
netcoredbg is an open-source debugger for .NET Core applications. The most popular one, vsdbg, is a binary blob distributed with the Visual Studio Code C# extension. However, [vsdbg's license](https://visualstudio.microsoft.com/license-terms/mt644895/) is incredibly restrictive. netcoredbg is a [drop-in replacement](https://github.com/VSCodium/vscodium/issues/82), and is also compatible with [other editors](https://github.com/puremourning/vimspector#supported-languages).

Note that the package here is... kind of awful. This is both because of the fact that netcoredbg uses CMake in pretty strange ways, and because packaging .NET Core applications in general is pretty bad at the moment.

Closes #103938

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/netcoredbg/create_deps.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+# Generate lockfiles in source checkout
+cd $CHECKOUT_PATH/src
+dotnet nuget locals all --clear
+dotnet restore -v normal --no-cache debug/netcoredbg/SymbolReader.csproj --use-lock-file -r linux-x64
+
+# Use the lockfiles to make a file with two columns: name and version number
+# for all possible package dependencies
+cd $SCRIPTDIR
+echo "" > all_versions.txt
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/create_deps.sh:24]:
```suggestion
touch all_versions.txt
```

File: pkgs/development/tools/misc/netcoredbg/create_deps.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -p python3 dotnet-sdk_3 -i bash
+
+# Run this script to generate deps.nix
+# ./create_deps.sh /path/to/netcoredbg/source/checkout
+
+SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+if [ -d "$1" ]; then
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/create_deps.sh:9]:
```suggestion
if [[ -d $1 ]]; then
```

File: pkgs/development/tools/misc/netcoredbg/create_deps.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+# Use the lockfiles to make a file with two columns: name and version number
+# for all possible package dependencies
+cd $SCRIPTDIR
+echo "" > all_versions.txt
+for lockfile in $(find "$CHECKOUT_PATH" -name packages.lock.json); do
+    echo "Processing lockfile $lockfile"
+    python ./process_lockfile.py "$lockfile" >> all_versions.txt
+done
+cat all_versions.txt | sed '/^$/d' | sort | uniq > tmp
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/create_deps.sh:29]:
```suggestion
cat all_versions.txt | sed '/^$/d' | sort -u > tmp
```

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+    pushd src
+    nuget sources Disable -Name "nuget.org"
+    dotnet restore --source ${nugetSource}/lib debug/netcoredbg/SymbolReader.csproj
+    popd
+
+    pushd src/debug/netcoredbg
+    dotnet build --no-restore -c Release SymbolReader.csproj
+    popd
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/default.nix:84]:
```suggestion
      mkdir -p $out/lib

      export HOME=$(mktemp -d)
      export DOTNET_CLI_TELEMETRY_OPTOUT=1
      export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1

      pushd src
      nuget sources Disable -Name "nuget.org"
      dotnet restore --source ${nugetSource}/lib debug/netcoredbg/SymbolReader.csproj
      popd

      pushd src/debug/netcoredbg
      dotnet build --no-restore -c Release SymbolReader.csproj
  popd
```

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    '';
+
+    installPhase = ''
+    pushd src/debug/netcoredbg
+    dotnet publish --no-restore -o $out/lib -c Release SymbolReader.csproj
+    popd
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/default.nix:90]:
```suggestion
    pushd src/debug/netcoredbg
        dotnet publish --no-restore -o $out/lib -c Release SymbolReader.csproj
      popd
```

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  inherit pname version;
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  buildCommand = ''
+  mkdir -p $out
+  cp ${unmanaged}/* $out
+  cp ${managed}/lib/* $out
+  makeWrapper $out/netcoredbg $out/bin/netcoredbg
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/default.nix:103]:
```suggestion
    mkdir -p $out
    cp ${unmanaged}/* $out
    cp ${managed}/lib/* $out
    makeWrapper $out/netcoredbg $out/bin/netcoredbg
  '';
```

File: pkgs/development/tools/misc/netcoredbg/format-deps.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -p gawk nix -i bash
+
+# Retrieve sha256 hashes for each dependency in and format fetchNuGet calls
+echo "" > deps.nix
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/format-deps.sh:5]:
```suggestion
touch deps.nix
```

File: pkgs/development/tools/misc/netcoredbg/format-deps.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+in [
+EOL
+IFS=''
+while read line; do
+  name=$(echo $line | awk '{print $1}')
+  version=$(echo $line | awk '{print $2}')
+  sha256=$(nix-prefetch-url "$urlbase/$name/$version" 2>/dev/null)
+
+  if [ -n "$sha256" ]; then
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/netcoredbg/format-deps.sh:26]:
```suggestion
  if [[ -n $sha256 ]]; then
```

**Comment by felschr** [comment]
I'm looking forward to this addition. Great work so far!
Are you planning to finish the PR?

**Comment by leo60228** [comment]
I was having some trouble getting it to build with .NET 5. I'm going to try again soon, maybe later today.

File: pkgs/development/tools/misc/netcoredbg/default.nix (hunk 4 above)
Comment (leo60228) [reply pkgs/development/tools/misc/netcoredbg/default.nix]: GitHub's UI is presenting this strangely and I can't tell what the proposed change here is.

**Comment by leo60228** [comment]
I've updated this to the latest netcoredbg version, .NET 5, and addressed the review.

**Comment by felschr** [comment]
You could get the runtime version with something like this, if that's the problem:
```sh
dotnet --list-runtimes | grep -Po '^Microsoft.NETCore.App \K.*?(?= )'
```

**Review by IvarWithoutBones — CHANGES_REQUESTED** [review]
Just a few minor comments, but this mostly looks good :)

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+      export HOME=$(mktemp -d)
+      export DOTNET_CLI_TELEMETRY_OPTOUT=1
+      export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+    '';
+
+    hardeningDisable = [ "strictoverflow" ];
+
+    preConfigure = ''
+      dotnetVersion="$(${dotnet-sdk}/bin/dotnet --list-runtimes | grep -Po '^Microsoft.NETCore.App \K.*?(?= )')"
+      cmakeFlags="$cmakeFlags -DDBGSHIM_RUNTIME_DIR=${dotnet-sdk}/shared/Microsoft.NETCore.App/$dotnetVersion"
```
Comment (IvarWithoutBones) [inline pkgs/development/tools/misc/netcoredbg/default.nix:40]: This looks a bit cleaner to me:
```suggestion
      cmakeFlagsArray+=(
        "-DDBGSHIM_RUNTIME_DIR=${dotnet-sdk}/shared/Microsoft.NETCore.App/$dotnetVersion"
      )

**Review by IvarWithoutBones — CHANGES_REQUESTED** [review]
Just a few more comments :)

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    projectFile = "src/managed/ManagedPart.csproj";
+    nugetDeps = ./deps.nix;
+
+    executables = [ ];
+  };
+in
+stdenvNoCC.mkDerivation {
+  inherit pname version;
+
+  nativeBuildInputs = [ makeWrapper ];
```
Comment (IvarWithoutBones) [inline pkgs/development/tools/misc/netcoredbg/default.nix:64]: This isn't used anymore
```suggestion
```

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    cp ${managed}/lib/netcoredbg/* $out/share/netcoredbg
+    ln -s $out/share/netcoredbg/netcoredbg $out/bin/netcoredbg
+  '';
+
+  meta = with lib; {
+    description = "Managed code debugger with MI interface for CoreCLR";
+    homepage = "https://github.com/Samsung/netcoredbg";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.unix;
+    maintainers = [ lib.maintainers.leo60228 ];
```
Comment (IvarWithoutBones) [inline pkgs/development/tools/misc/netcoredbg/default.nix:78]: As we used `with lib;` before, it's already in scope:
```suggestion
    license = licenses.mit;
    platforms = platforms.unix;
    maintainers = [ maintainers.leo60228 ];
```

File: pkgs/development/tools/misc/netcoredbg/default.nix [hunk 12]
```diff
@@ -0,0 +1,80 @@
+{ lib, clangStdenv, stdenvNoCC, makeWrapper, cmake, fetchFromGitHub, dotnetCorePackages, buildDotnetModule }:
```
Comment (IvarWithoutBones) [inline pkgs/development/tools/misc/netcoredbg/default.nix:1]: `makeWrapper` can be removed here as well.

**Review by IvarWithoutBones — CHANGES_REQUESTED** [review]
Some final nits, though functionality wise this looks good to me :+1: 

Result of `nixpkgs-review` run on x86_64-linux
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>netcoredbg</li>
  </ul>
</details>

The binary also works perfectly as expected :)

**Review by felschr — APPROVED** [review]

**Review by IvarWithoutBones — APPROVED** [review]
LGTM, thanks for the work!

**Pull request merged**